### PR TITLE
APEXMALHAR-2017

### PR DESCRIPTION
**Pull request by prasannapramod**
@tweise @PramodSSImmaneni please review. Originally submitted by pramod. Multiple tests were failing with latest, so fixed tests.

I had accidentally pushed wrong branch in the earlier pull request and I have since closed it. Please review this one instead. 

@tweise regarding your comments on earlier pull request

a) Extending both the interfaces is for japicmp 
b) Regarding getting the file output operator changes reviewed, Pramod had already asked Chandni in his original pull request and she was ok with the changes, I will ask her again.

### CODE REVIEW

File: kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java [hunk 1]
```diff
@@ -194,6 +194,11 @@ public void deactivate()
   }
 
   @Override
+  public void beforeCheckpoint(long windowId)
```
Comment (tweise) [inline kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: Why this change?

**Review by tweise — CHANGES_REQUESTED** [review]

**Comment by tweise** [comment]
If we can configure a japicmp exclusion so that we can get rid of "implements CheckpointListener" it would be great. If not it is fine also.

Other changes LGTM.

File: kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java (hunk 1 above)
Comment (pramodin) [reply kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: For some reason I am not seeing this change when I merge this pull request into my local branch. @tweise do you have a fresh merge from this pull request?

Comment (tweise) [reply kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: It's in the first commit.

**Comment by prasannapramod** [comment]
@chandnisingh can you look at the file output operator changes

File: kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java (hunk 1 above)
Comment (prasannapramod) [reply kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: I see, I undid that change in my second commit. @PramodSSImmaneni can you fix the first commit so it is cleaner.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 2]
```diff
@@ -260,6 +260,8 @@
   private Long expireStreamAfterAccessMillis;
   private final Set<String> filesWithOpenStreams;
 
+  private boolean initializeContext;
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: should this be transient?

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 3]
```diff
@@ -923,32 +925,23 @@ protected String getPartFileName(String fileName, int part)
   @Override
   public void beginWindow(long windowId)
   {
-    try {
-      Map<String, FSFilterStreamContext> openStreams = streamsCache.asMap();
-      for (FSFilterStreamContext streamContext : openStreams.values()) {
-        streamContext.initializeContext();
+    if (initializeContext) {
+      try {
+        Map<String, FSFilterStreamContext> openStreams = streamsCache.asMap();
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java:976]: Why do we need to initialize contexts after a checkpoint? Please add a comment so it is clear.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 4]
```diff
@@ -1196,6 +1189,24 @@ public void close() throws IOException
...
   @Override
+  public void beforeCheckpoint(long l)
+  {
+    try {
+      Map<String, FSFilterStreamContext> openStreams = streamsCache.asMap();
+      for (FSFilterStreamContext streamContext: openStreams.values()) {
+        long start = System.currentTimeMillis();
+        streamContext.finalizeContext();
+        totalWritingTime += System.currentTimeMillis() - start;
+        //streamContext.resetFilter();
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: Nitpick: commented out code need not be copied over

File: library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java [hunk 5]
```diff
@@ -243,8 +243,11 @@ private void populateFile(String fileName, String contents) throws IOException
    * @param writer The writer to checkpoint.
    * @return new writer.
    */
-  public static AbstractFileOutputOperator checkpoint(AbstractFileOutputOperator writer)
+  public static AbstractFileOutputOperator checkpoint(AbstractFileOutputOperator writer, long windowId)
   {
+    if (windowId >= 0) {
```
Comment (otterc) [inline library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java]: Nitpick: windowId > Stateless.WindowId

File: library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java [hunk 6]
```diff
@@ -862,8 +868,8 @@ public void testSingleRollingFileFailedWrite1()
     writer.input.put(4);
     writer.endWindow();
 
-    AbstractFileOutputOperator checkPointWriter = checkpoint(writer);
-    AbstractFileOutputOperator checkPointWriter1 = checkpoint(writer);
+    AbstractFileOutputOperator checkPointWriter = checkpoint(writer, 0);
+    AbstractFileOutputOperator checkPointWriter1 = checkpoint(writer, -1);
```
Comment (otterc) [inline library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java]: Nitpick: instead of -1, Stateless.WindowId

File: library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java [hunk 7]
```diff
@@ -1756,11 +1766,14 @@ public void testChainFilters() throws NoSuchAlgorithmException, IOException
       for (int j = 0; j < 1000; ++j) {
         writer.input.put(i);
       }
-      writer.endWindow();
-      evenOffsets.add(evenCounterContext.getCounter());
-      oddOffsets.add(oddCounterContext.getCounter());
-      //evenOffsets.add(evenFile.length());
-      //oddOffsets.add(oddFile.length());
+      //writer.endWindow();
```
Comment (otterc) [inline library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java]: why commenting endWindow()? engine calls endWindow() before `beforeCheckpoint()` callback so there isn't any need to comment this out.

File: library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java [hunk 8]
```diff
@@ -1756,11 +1766,14 @@ public void testChainFilters() throws NoSuchAlgorithmException, IOException
...
+      if ((i % 2) == 1) {
+        writer.beforeCheckpoint(i);
+        evenOffsets.add(evenCounterContext.getCounter());
+        oddOffsets.add(oddCounterContext.getCounter());
+        //evenOffsets.add(evenFile.length());
+        //oddOffsets.add(oddFile.length());
```
Comment (otterc) [inline library/src/test/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperatorTest.java]: Nitpick: please remove commented out code

**Review by otterc — COMMENTED** [review]
Besides minor comments/nitpicks, LGTM

**Comment by tweise** [comment]
@prasannapramod  any update on this?

**Comment by prasannapramod** [comment]
@tweise Hive operator tests are failing. I am looking into it.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java (hunk 3 above)
Comment (prasannapramod) [reply library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java:976]: @chandnisingh because at checkpoint, all the filter state needs to be flushed to the disk, not all filters allow a flush option, so the filters have to be closed and reopened. If no filter being is being used then it is a essentially a noop as the underlying FSDataOutputStream is not being closed in this operation.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java (hunk 2 above)
Comment (prasannapramod) [reply library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: @chandnisingh yes this should be transient. i will change it.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java (hunk 3 above)
Comment (otterc) [reply library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java:976]: Can this be documented with the code?

Comment (prasannapramod) [reply library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java:976]: @chandnisingh sure.

**Comment by prasannapramod** [comment]
@tweise @chandnisingh @PramodSSImmaneni addressed chandni's comments, fixed Hive tests, squashed down to 2 commits, one from pramod & another from me and rebased to latest master. Let me know if there is anything else.

**Comment by pramodin** [comment]
LGTM. I can merge but since I also have a commit in this, if either of you want to merge please go ahead.
